lib/reversetunnel/srv.go

```diff
@@ -636,6 +544,49 @@ func (s *server) Shutdown(ctx context.Context) error {
 	return trace.Wrap(err)
 }
 
+// HandleRequest processes global out-of-band requests. Global out-of-band
+// requests are processed in order (this way the originator knows which
+// request we are responding to). If Teleport does not support the request
+// type or an error occurs while processing that request Teleport will reply
+// req.Reply(false, nil).
+//
+// For more details: https://tools.ietf.org/html/rfc4254.html#page-4
+func (s *server) HandleRequest(ctx context.Context, r *ssh.Request) {
+	switch r.Type {
+	case teleport.NetconfigRequest:
+		s.handleNetConfigRequest(ctx, r)
+	default:
+		if err := r.Reply(false, nil); err != nil {
+			s.log.Warnf("Failed to reply to %q request: %v", r.Type, err)
+		}
+		s.log.Debugf("Discarding %q global request: %+v", r.Type, r)
+	}
+}
+
+// handleNetConfigRequest replies to a ssh.Request with the cluster  networking config.
+func (s *server) handleNetConfigRequest(ctx context.Context, req *ssh.Request) {
+	if !req.WantReply {
+		return
+	}
+
+	// On error, reply with true to indicate that the server is able to handle the request
+	// but a server side error occurred, this lets the agent differentiate between a server
+	// that does not support the request and a transient error.
+	netconfig, err := s.localAccessPoint.GetClusterNetworkingConfig(ctx)
+	if err != nil {
+		req.Reply(true, []byte(fmt.Sprintf("failed to get cluster networking config: %v", err)))
+	}
+	payload, err := json.Marshal(netconfig)
+	if err != nil {
+		req.Reply(true, []byte(fmt.Sprintf("failed to marshal cluster networking config: %v", err)))
```

Comment:
I really don't like relying on the fact that an error string doesn't unmarshal as a json object to do error detection on the other end. Can we use the `ok` bool of the reply to the request? The payload still goes through even if a request in NAKed.

Reply:
What do you think of a wrapper type to contain the error like 
```golang
type NetconfigWithError struct {
  ClusterNetworkingConfig types.ClusterNetworkingConfig
  Error *string
}
```

This makes it clear the request was not rejected and allows the client to check if the error is not nil.

Reply:
Why does it matter if the request was "not rejected"?

The agent itself replies with `Reply(r, true, []byte(version)` or `Reply(r, false, []byte("Failed to retrieve auth version"))` to a version request through a global request, I'd like to not add yet another standard for message passing through SSH.

Reply:
I suppose its not absolutely necessary. However the point was to allow the agentpool to differentiate between the request failing and the request not being supported. If the request failed this time around that means we can retry later but if the request is rejected (or in other words not supported) that means the agent pool should hit the /webapi/find endpoint and cobble together a netconfig with some defaults like using  the `agent_mesh` tunnel strategy. This helps maintain backwards compatibility.

Currently when the agentpool fails to retrieve a new netconfig it stops trying to connect new agents until it is able to retrieve the netconfig. If we know the request is not supported we can fallback to the default netconfig and continue connecting agents, if we know the request failed server side and can be retried we can retry.

Technically we know that if we can open the reverse tunnel v2 channel we know the netconfig request is supported and can be retried but it feels a kind of brittle to make that assumption.

Reply:
Updated here [4bfe263](https://github.com/gravitational/teleport/pull/24702/commits/4bfe2630fe272c559e66e63397846acec51aa2c6)